user_tools/src/spark_rapids_tools/tools/cluster_config_recommender.py

```diff
@@ -0,0 +1,190 @@
+
+import pandas as pd
+
+from spark_rapids_pytools.cloud_api.sp_types import ClusterBase
+from spark_rapids_pytools.common.sys_storage import FSUtil
+from spark_rapids_pytools.rapids.tool_ctxt import ToolContext
+
+
+@dataclass
+class ClusterRecommendationInfo:
+    """
+    Dataclass to hold the recommended cluster and the qualified node recommendation.
+    """
+    source_cluster_config: dict = field(default_factory=dict)
+    recommended_cluster_config: dict = field(default_factory=dict)
+    qualified_node_recommendation: str = 'Not Available'
+
+    def to_dict(self) -> dict:
+        return {
+            'Source Cluster': self.source_cluster_config,
+            'Recommended Cluster': self.recommended_cluster_config,
+            'Qualified Node Recommendation': self.qualified_node_recommendation
+        }
+
+    @classmethod
+    def get_default_dict(cls) -> dict:
+        return cls().to_dict()
+
+
+@dataclass
+class TuningRecommendationInfo:
+    """
+    Dataclass to hold the full cluster config recommendations and the GPU config recommendation
+    """
+    cluster_config_recommendations: pd.Series
+    gpu_config_recommendation: pd.Series
+
+    def to_dict(self) -> dict:
+        return {
+            'Full Cluster Config Recommendations*': self.cluster_config_recommendations.to_list(),
+            'GPU Config Recommendation Breakdown*': self.gpu_config_recommendation.to_list()
+        }
+
+    @classmethod
+    def get_default(cls, missing_msg: str, num_apps: int = 0) -> 'TuningRecommendationInfo':
+        default_series = pd.Series([missing_msg] * num_apps)
+        return cls(cluster_config_recommendations=default_series, gpu_config_recommendation=default_series)
+
+
+@dataclass
+class ClusterConfigRecommender:
+    """
+    Class for recommending cluster shape and tuning configurations for processed apps.
+    """
+    ctxt: ToolContext = field(default=None, init=True)
+
+    @classmethod
+    def _get_instance_type_conversion(cls, cpu_cluster: ClusterBase,
+                                      gpu_cluster: ClusterBase) -> Optional[ClusterRecommendationInfo]:
+        """
+        Helper method to determine the conversion summary between CPU and GPU instance types.
+        Generate the cluster shape recommendation as:
+        {
+          'Source Cluster': {'driverInstance': 'm6.xlarge', 'executorInstance': 'm6.xlarge', 'numExecutors': 2 }
+          'Recommended Cluster': {'driverInstance': 'm6.xlarge', 'executorInstance': 'g5.2xlarge', 'numExecutors': 2 }
+          'Qualified Node Recommendation': 'm6.xlarge to g5.2xlarge'
+        }
+        """
+        if not cpu_cluster:
+            return None
+
+        cpu_instance_type = cpu_cluster.get_worker_node().instance_type
+        recommended_cluster = cpu_cluster
+        conversion_str = cpu_instance_type
+        if gpu_cluster:
+            gpu_instance_type = gpu_cluster.get_worker_node().instance_type
+            if cpu_instance_type != gpu_instance_type:
```

Comment:
Curious when will this condition be false? Does the gpu_worker_node return cpu_instance type? And if this condition is false 'Qualified Node Recommendation' will be a CPU instance?

Reply:
This condition can be false for dataproc platform. In dataproc, `n1-standard-xx` can be both CPU and GPU instance.

Reply:
Thanks @parthosa ! Still going through the code.